SimDataFormats/GeneratorProducts/src/HepMC3Product.cc

```diff
@@ -0,0 +1,143 @@
+
+using namespace edm;
+using namespace std;
+
+HepMC3Product::HepMC3Product(HepMC3::GenEvent* evt)
+    : evt_(evt), isVtxGenApplied_(false), isVtxBoostApplied_(false), isPBoostApplied_(false) {}
+
+HepMC3Product::~HepMC3Product() {
+  delete evt_;
+  evt_ = nullptr;
+  isVtxGenApplied_ = false;
+  isVtxBoostApplied_ = false;
+  isPBoostApplied_ = false;
+}
+
+void HepMC3Product::addHepMCData(HepMC3::GenEvent* evt) {
+  //  evt->print();
+  // cout <<sizeof (evt)  <<"  " << sizeof ( HepMC::GenEvent)   << endl;
+  evt_ = evt;
+
+  // same story about vertex smearing - GenEvent won't know it...
+  // in fact, would be better to implement CmsGenEvent...
+}
+
+void HepMC3Product::applyVtxGen(HepMC3::FourVector const& vtxShift) {
+  //std::cout<< " applyVtxGen called " << isVtxGenApplied_ << endl;
+  //fTimeOffset = 0;
+  if (isVtxGenApplied())
+    return;
+  evt_->shift_position_by(vtxShift);
+  isVtxGenApplied_ = true;
+  return;
+}
+
+void HepMC3Product::boostToLab(TMatrixD const* lorentz, std::string const& type) {
+  //std::cout << "from boostToLab:" << std::endl;
+
+  if (lorentz == nullptr) {
+    //std::cout << " lorentz = 0 " << std::endl;
+    return;
+  }
+
+  //lorentz->Print();
+
+  TMatrixD tmplorentz(*lorentz);
+  //tmplorentz.Print();
+
+  if (type == "vertex") {
+    if (isVtxBoostApplied()) {
+      //std::cout << " isVtxBoostApplied true " << std::endl;
+      return;
+    }
+
+    for (const HepMC3::GenVertexPtr& vt : evt_->vertices()) {
+      // change basis to lorentz boost definition: (t,x,z,y)
+      TMatrixD p4(4, 1);
+      p4(0, 0) = vt->position().t();
+      p4(1, 0) = vt->position().x();
+      p4(2, 0) = vt->position().z();
+      p4(3, 0) = vt->position().y();
+
+      TMatrixD p4lab(4, 1);
+      p4lab = tmplorentz * p4;
+      //std::cout << " vertex lorentz: " << p4lab(1,0) << " " << p4lab(3,0) << " " << p4lab(2,0) << std::endl;
+      vt->set_position(HepMC3::FourVector(p4lab(1, 0), p4lab(3, 0), p4lab(2, 0), p4lab(0, 0)));
+    }
+
+    isVtxBoostApplied_ = true;
+  } else if (type == "momentum") {
+    if (isPBoostApplied()) {
+      //std::cout << " isPBoostApplied true " << std::endl;
+      return;
+    }
+
+    for (const HepMC3::GenParticlePtr& part : evt_->particles()) {
+      // change basis to lorentz boost definition: (E,Px,Pz,Py)
+      TMatrixD p4(4, 1);
+      p4(0, 0) = part->momentum().e();
+      p4(1, 0) = part->momentum().x();
+      p4(2, 0) = part->momentum().z();
+      p4(3, 0) = part->momentum().y();
+
+      TMatrixD p4lab(4, 1);
+      p4lab = tmplorentz * p4;
+      //std::cout << " momentum lorentz: " << p4lab(1,0) << " " << p4lab(3,0) << " " << p4lab(2,0) << std::endl;
+      part->set_momentum(HepMC3::FourVector(p4lab(1, 0), p4lab(3, 0), p4lab(2, 0), p4lab(0, 0)));
+    }
+
+    isPBoostApplied_ = true;
+  } else {
+    std::cout << " no type found for boostToLab(std::string), options are vertex or momentum" << std::endl;
```

Comment:
couts should be moved to LogPrint (cout is not thread safe)

Reply:
@mkirsano , you may comment out cout for the time being and later add logVerbatim or logWarning if needed.